subsys/mgmt/mcumgr/grp/settings_mgmt/src/settings_mgmt.c

```diff
@@ -0,0 +1,535 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(settings_mgmt);
+
+/**
+ * Command handler: settings read
+ */
+static int settings_mgmt_read(struct smp_streamer *ctxt)
+{
+	int rc;
+	zcbor_state_t *zse = ctxt->writer->zs;
+	zcbor_state_t *zsd = ctxt->reader->zs;
+	bool ok;
+	size_t decoded;
+	struct zcbor_string key = { 0 };
+	uint32_t max_size = CONFIG_MCUMGR_GRP_SETTINGS_VALUE_LEN;
+	bool limited_size = false;
+
+#ifdef CONFIG_MCUMGR_GRP_SETTINGS_BUFFER_TYPE_HEAP
+	char *key_name = NULL;
+	uint8_t *data = NULL;
+#else
+	char key_name[CONFIG_MCUMGR_GRP_SETTINGS_NAME_LEN] = { 0 };
+	uint8_t data[CONFIG_MCUMGR_GRP_SETTINGS_VALUE_LEN] = { 0 };
+#endif
+
+	struct zcbor_map_decode_key_val settings_read_decode[] = {
+		ZCBOR_MAP_DECODE_KEY_DECODER("name", zcbor_tstr_decode, &key),
+		ZCBOR_MAP_DECODE_KEY_DECODER("max_size", zcbor_uint32_decode, &max_size),
+	};
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+	struct settings_mgmt_access settings_access_data = {
+		.access = SETTINGS_ACCESS_READ,
+		.name = key_name,
+	};
+
+	enum mgmt_cb_return status;
+	int32_t ret_rc;
+	uint16_t ret_group;
+#endif
+
+	ok = zcbor_map_decode_bulk(zsd, settings_read_decode, ARRAY_SIZE(settings_read_decode),
+				   &decoded) == 0;
+
+	if (!ok || key.len == 0) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Check if the length of the user supplied key is too large for the configuration
+	 * to allow and return an error if so
+	 */
+	if (key.len >= CONFIG_MCUMGR_GRP_SETTINGS_NAME_LEN) {
+		ok = smp_add_cmd_ret(zse, MGMT_GROUP_ID_SETTINGS,
+				     SETTINGS_MGMT_RET_RC_KEY_TOO_LONG);
+		goto end;
+	}
+
+	if (max_size > CONFIG_MCUMGR_GRP_SETTINGS_VALUE_LEN) {
+		max_size = CONFIG_MCUMGR_GRP_SETTINGS_VALUE_LEN;
+		limited_size = true;
+	}
+
+#ifdef CONFIG_MCUMGR_GRP_SETTINGS_BUFFER_TYPE_HEAP
+	key_name = (char *)malloc(key.len + 1);
+	data = (uint8_t *)malloc(max_size);
+
+	if (data == NULL || key_name == NULL) {
+		if (key_name != NULL) {
+			free(key_name);
+		}
+
+		return MGMT_ERR_ENOMEM;
+	}
+#endif
+
+	memcpy(key_name, key.value, key.len);
+	key_name[key.len] = 0;
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+	/* Send request to application to check if access should be allowed or not */
+	status = mgmt_callback_notify(MGMT_EVT_OP_SETTINGS_MGMT_ACCESS, &settings_access_data,
+				      sizeof(settings_access_data), &ret_rc, &ret_group);
+
+	if (status != MGMT_CB_OK) {
+		bool ok;
+
+		if (status == MGMT_CB_ERROR_RC) {
+			return ret_rc;
+		}
+
+		ok = smp_add_cmd_ret(zse, ret_group, (uint16_t)ret_rc);
+		goto end;
+	}
+#endif
+
+	/* Read the settings key */
+	rc = settings_runtime_get(key_name, data, max_size);
+
+	if (rc < 0) {
+		if (rc == -EINVAL) {
+			rc = SETTINGS_MGMT_RET_RC_KEY_NOT_FOUND;
+		} else if (rc == -ENOTSUP) {
+			rc = SETTINGS_MGMT_RET_RC_READ_NOT_SUPPORTED;
+		} else {
+			rc = SETTINGS_MGMT_RET_RC_UNKNOWN;
+		}
+
+		ok = smp_add_cmd_ret(zse, MGMT_GROUP_ID_SETTINGS, (uint16_t)rc);
+		goto end;
+	}
```

Comment:
OMG, I have completely forgotten that this thing will not return info that it has returned less data then there was stored, and there is now way to get info on the length of value.

Reply:
Indeed, I do not really like the settings subsystem, it is janky, really hard to setup and use and cannot provide up-front information on keys